util/ss/ss.man
'\" t
.\" Run through tbl first
.\" **********************************************************************
.\"              Copyright (c) 1992 by Arthur E. Mulder.
.\"
.\"         (See the end of this file for a copyright notice)
.\" **********************************************************************
.\" * %M%
.\" * ss    :       A SpreadSheet Program
.\" *
.\" * Art's Spreadsheet program.        Art Mulder ( art@cs.ualberta.ca )
.\" * University of Alberta, Department of Computing Science.
.\" **********************************************************************
.\" * Main Documentation (Man Page)
.\" **********************************************************************
.\"
.\" USAGE: tbl <this file> | nroff -man 
.\"
.\" WARNING:
.\"   1) The string "pname" is converted to the true program name
.\"      by the makefile, throughout this document.
.\"   2) The string "#LIBDIR#" is converted to the value of the variable
.\"	 LIBDIR from the Makefile.
.\"   3) The string "#REVISION#" is converted to the revision number of
.\"      of the package, as stripped from 'version.c', by 'torev'.
.\"
.\" CONVENTIONS:
.\" - pname italicized and never uppercased (it's a proper name).
.\" - Refer to lists of commands in the same order as introduced.
.\" - Command and function names bold when introduced, italicized in all
.\"   other places if possible, or in `` '' if not.
.\" - Cell names italicized except when used in expressions; row numbers
.\"   and column names not italicized.
.\" - Use `` '' rather than " " except referring to literal input or output.
.\" - TPs use default indent except for function names, then 18.
.\" - Smallify uppercase strings.
.\" - Avoid passive voice and third person.
.\" $Revision: 1.1 $
.\" **********************************************************************
.\"
.TH PNAME\ #REVISION#\  L
.SH NAME
pname \- simple spreadsheet program
.SH SYNOPSIS
.B pname
[
.B -c
]
[
.B -m
]
[
.B -r
]
[
.B -x
]
[
.B -C
]
[
.B -R
]
[
.I file
]
.\"
.\" **********************************************************************
.SH DESCRIPTION
.\" ***********
.I pname
is a spreadsheet program (big surprise!).  
.LP
It is not another Lotus 1-2-3(tm) or Excel(tm) or etc spreadsheet
program that consumes megabytes of disk space and is full of whiz-bang
graphics and fonts and so on and so forth that the average user (IMHO)
needs and uses only infrequently.
.LP
.I pname
is a character based program that should run on any standard ASCII
terminal.  It is based on rectangular tables much like a financial
spreadsheet.  When invoked it presents you with a table organized as
rows and columns of cells \- just like any ``normal'' spreadsheet
program.  
.\" ** .LP
.\" ** For a online tutorial, type the command:
.\" ** .IP
.\" ** pname #LIBDIR#/tutorial.pname
.\" ** .LP
.\" ** To print a quick reference card, type the command:
.\" ** .IP
.\" ** pnameqref | [your_printer_commmand]
.\"
.\" **********************************************************************
.SH OPTIONS
.\" ***********
.TP
.B \-c
Start the program with the recalculation being done in column order.
.\" ----------------------------------------------------------------------
.TP
.B \-m
Start the program with automatic recalculation disabled.  The
spreadsheet will be recalculated only when the recalc command (see
below) is used.
.\" ----------------------------------------------------------------------
.TP
.B \-r
Start the program with the recalculation being done in row order
(default option).
.\" ----------------------------------------------------------------------
.TP
.B \-x
Cause the data files to be decrypted/encrypted when read/written.  (This
option may not be available at your site, depending on the capabilities
of your computer system, check with your system administrator.)
.\" ----------------------------------------------------------------------
.TP
.B \-R
Start the program with automatic newline action set to increment the
row (see below).
.\" ----------------------------------------------------------------------
.TP
.B \-C
Start the program with automatic newline action set to increment the
column (see below).
.\" ----------------------------------------------------------------------
.TP
.B file
If invoked without a
.I file
argument, the table is initially empty.  Otherwise
.I file
is read in.
.\" ----------------------------------------------------------------------
.PP
All of these options can also be changed from within
.IR pname .
Those changes will then be saved when the current spreadsheet is saved,
and reloaded when that spreadsheet is read in from disk.  Options
specified when
.I pname
is invoked override options saved in the data file.
.\"
.\"
**********************************************************************
.SH INTRODUCTION
.\" ***********
.I pname
is based upon the well-known (?)  public domain spreadsheet program
.RB `` sc '',
version 6.19.
.LP
.I pname
is an ``un-moded'' program.  It does not have separate input and output
modes that you toggle between.  To enter numbers into your
spreadsheet, go ahead and start typing.  You can enter text the same
way, just do it (more detail on these operations below).
.LP
If your keyboard has Arrow keys and/or Page-Up and Page-Down keys,
they should work as you expect them too.  If they don't, talk to
whomever installed this program on your system, it could be a simple
terminfo/termcap problem.  If your keyboard doesn't have cursor
movement keys, relax.  When you use
.I pname 
you can also use the same cursor movement commands as emacs  (more
on that below, also).
.LP
There are a wide selection of financial and other ``standard''
spreadsheet functions that you can use.  They are listed and explained
in detail below.  (Wow, sure must be a lot of stuff below!)
.LP
Finally, instead of forcing you to memorize bazillions of different
commands, 
.I pname
places most spreadsheet operations (Save, Move, etc) in menus.  These
menus are probably vaguely familiar to you, since they were inspired
by a certain well-known DOS spreadsheet program.  Go ahead, hit the
``/'' key.  You probably will figure many things out without even
reading on in this man page.
.LP
However, you probably should read on, because there's lot's to learn.
I'd also hate to have typed all this in for nothing!  :-)
.LP
NOTE: This man page assumes that the reader has a general familiarity
with spreadsheet programs.
.\" ----------------------------------------------------------------------
.SS Screen Layout
.\" ----------------------------------------------------------------------
The screen is divided into four regions.  The top line is for entering
commands/data and displaying cell values.  The second line is for messages
from
.IR pname .
The third line and the first four columns show the column and row
numbers, from which are derived cell addresses, e.g.
.I A0
for the cell in column A, row 0.  Note that column names are
case-insensitive: Entering
.I A0
is equivalent to
.IR a0 .
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.PP
The rest of the screen forms a window looking at a portion of the
table.  The total number of table rows and columns displayed is set
by
.IR curses (3)
and may be overridden by setting the LINES and COLUMNS environment
variables, respectively.
.PP
.RS
Note to X-Windows users; Yes you can resize your terminal windows.
The next time a command is processed, or the screen is redrawn,
.I pname
will figure things out and adjust itself.
.RE
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.PP
The screen has two cursors:  a cell cursor, indicated by a highlighted
cell and a ``<'' on the screen, and a character cursor, indicated by
the terminal's hardware cursor.  The cell and character cursors are
often the same.  They differ when you type a command on the top line.
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.PP
If a cell's numeric value is wider than the column width the cell is
filled with asterisks.  If a cell's label string is wider than the
column width, it is truncated at the start of the next non-blank cell
in the row, if any.
.\" ----------------------------------------------------------------------
.SH How Input Is Processed
.\" ----------------------------------------------------------------------
Let's assume that you have just started
.IR pname ,
so that you are at the ``top-level'' of input processing.  You enter
a keystroke, a number or letter, arrow key or function key.  The
program reads your input and processes it as follows:
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP
.B Control-key, Arrow key, or Function key
The appropriate function is performed  (ie: Move the cursor to
the next row).  See the section below entitled
.BR COMMANDS .
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP
.B `/'
The main top-level menu is displayed and the program enters
``menu-mode''.  See the section below entitled 
.BR MENUS .
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.\" ** .IP `?'
.\" ** Pop-up help is invoked.  CURRENTLY DISABLED.
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP
.B `0-9',`-',`.',`+',`@'
It is assumed that you are entering a number (or a function, in the
case of `@'), into the current cell.
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP
.B `='
This is retained for compatibility with ``sc'' .  It signals that you
wish to enter a number/function in the current cell.
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP
.B `"',`>',`<'
These characters indicate that you want to enter a centered,
right-justified, or left-justified string into the cell.
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP
.B Default:
The entry of any other character is taken to be a signal that you wish
to enter a left-justified string.
.\"
.\" **********************************************************************
.SH MENUS
.\" ***********
The main (top-level) menu is invoked with the `/' key.  
.LP
Menus are displayed on the top line of the screen.  Each menu item is
usually one (hopefully self-explanatory) word.  Each menu item begins
with a single highlighted letter.  By typing the key corresponding to
the letter \- case is unimportant \- you select that menu item.
.LP 
I tried to have all the highlighted letters match the first letter of
their menu item, but unfortunately there had to be a few exceptions.
Watch for those!
.LP
By typing a <Return> or a <Space> you will abort out of any menu (no
matter how deep you are into sub-menu's) back to the top-level of the
program.
.\" ----------------------------------------------------------------------
.SS Main Menu
.BR F :File 
.BR E :Edit
.BR C :Cell
.BR R :Row
.BR O :Column
.BR M :Misc
.BR A :Macro
.BR Q :Quit
.LP
With the exception of 
.BR Quit ,
all menu items on the Main Menu invoke sub-menus.  See below for
explanations of those.
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP
.I Quit
Exit from the program.  If there have been any changes in the
spreadsheet since the last Load or Save command
.I pname
asks about saving your data before exiting.
.\" ----------------------------------------------------------------------
.SS File Sub-Menu
.BR N :New
.BR L :Load
.BR M :Merge
.BR S :Save
.BR A :Save As
.BR W :Write
txt
.BR T :Tbl
.RI save( mode )
.BR Q :Quit
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP
.I New
Erase the current spreadsheet from memory and start fresh with an
empty table.
.I (Unimplemented)
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP
.I Load
Load (``get'') a new spreadsheet file from disk.  Enter the name of a
spreadsheet file to load, followed by <CR>.  If encryption is enabled,
the file is decrypted before it is loaded into the spreadsheet.
.IP
NOTE: 
.I pname
files are (currently) identical to ``sc'' data files.
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP
.I Merge
Merge the spreadsheet from the named file into the current one.
Values and expressions defined in the named file are read into the
current spreadsheet, overwriting the existing entries at matching cell
locations.
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP
.I Save
Save (``put'') the current spreadsheet into a file.
If encryption is enabled, the file is encrypted before it is saved.
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP
.I Save As
Prompt the user to enter a filename, and then save (``put'') the
spreadsheet into a file with that name.  If a <CR> is entered, the
current default filename is used.  The newly entered filename becomes
the new default filename.  If encryption is enabled, the file is
encrypted before it is saved.
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP
.I Write Txt
Write a text-only listing of the current spreadsheet (No functions,
etc) into a file in a form that matches its appearance on the screen.
.IP
This differs from the ``Save'' commands in that files saved with those
commands are intended to be reloaded with ``Load'', while ``Write
txt'' produces a file for people to look at.  Hidden rows or columns
are not shown when the data is printed.
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP
.I Table Save
Write a text-only listing of the current spreadsheet to a file,
but include delimiters suitable for processing by the
.IR tbl ,
.IR LaTeX ,
.IR TeX ,
or 
.I FrameMaker
table processors.
.IP
The 
.I tblstyle
option (see the ``Settings'' item on the Misc menu) controls which
delimiters are output.  The delimiters are are a colon (:) for style
.I 0
(no style defined) or
.IR tbl ,
and an ampersand (&) for style
.I LaTeX
or
.IR TeX .
.IP
The current setting of the 
.I tblstyle 
option is displayed in parentheses after the ``Tbl Save'' entry of the
File menu.  A question mark (?) indicates that no style has been defined
(style 0).
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP
.I Quit
Exit from the program.  Identical to the Quit option on the main
top-most menu.  (Actually the Quit option on the main top-most
menu is identical to 
.I this
option, but that really is irrelevant).
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.PP
With the ``Save'', ``Write Text'', and ``Table Save'' commands,
it is possible to save a write a subset of the spreadsheet to the
output file.  To do that, a 
.I range
must be defined prior to invoking those commands.  See the section
below on Ranges, for instructions on defining a range.
.PP
With the ``Write Text'' and ``Table Save'' commands, if you try to
write to the last file used with the ``Load'' or ``Save/Save As''
commands, or the file specified on the command line when
.I pname
was invoked, you are asked to confirm that the (potentially) dangerous
operation is really what you want.
.PP
The four output commands, (``Save'', ``Save As'', ``Write Text'' and
``Table Save''), can pipe their (unencrypted only) output to a
program.  To use this feature, enter ``| program'' to the prompt
asking for a filename.  For example, to redirect the output of the
``Write Text'' command to the printer, you might enter ``| lpr -p''.
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.\" ** Hmmm, now that I've changed the way input is processed, this 
.\" ** next feature is no longer available, since I put the "filename"
.\" ** in quotes.  Can a workaround be discovered somehow?  Hmmm.
.\" 
.\" The filename can also be obtained from a cell's label string or string
.\" expression.  In this case, delete the leading " with the backspace key
.\" and enter a cell name such as
.\" .I a22
.\" instead.  If the resulting string starts with ``|'', the rest of the
.\" string is interpreted as a
.\" .SM UNIX
.\" command, as mentioned above.
.\"
.\" ----------------------------------------------------------------------
.SS Edit Sub-Menu
.BR C :Copy 
.BR E :Erase
.BR N :Name
.BR L :Lock
.BR U :Unlck
.BR F :Format
.BR I :Fill
.BR V :Valueize
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP
.I Copy 
Copy a source range to a destination range.  The source and
destination may be different sizes.  The result is always one or more
full copies of the source.  Copying a row to a row yields a row.
Copying a column to a column yields a column.  Copying a range to
anything yields a range.  Copying a row to a column or a column to a
row yields a range with as many copies of the source as there are
cells in the destination.  This command can be used to duplicate a
cell through an arbitrary range by making the source a single cell
range such as
.IR b20:b20 .
.IP
The Source Range must be defined before selecting this command.
You will be prompted to enter a destination range.
.IP
.I BUG: 
You MUST enter a range, and the range WILL be copied.  If you
want to abort this operation the only current solution is to select
your source range to also be your destination range.  Then there will
be a net effect of zero.  This should be addressed in the next release.
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP
.I Erase
Clear a range.  Cells cleared with this command may be recalled with
the ``Yank'' or ``Merge'' commands under the Row or Column menus.
(This is, of course, less than ideal \- this awkwardness is an example
of some of the problems that arose in converting from ``sc''.)
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP
.I Name
Invoke the Name sub-menu.
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP
.I Lock
This command will lock the current cell -- or a range of cells, if a
range of cells is specified.  Locking will make them immune to any
type of editing. A locked cell can't be changed in anyway until it is
unlocked.
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP
.I UnLock
In light of the previous command, the necessity of this command is
obvious.  This command will unlock a locked cell and make it editable.
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP
.I Format
Use this command to assign a value format string to a range of cells.
See the section ``Formatting Cell Contents'' for details.
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP
.I Fill
Fill a range with constant values starting with a given value and
increasing by a given increment.  Each row is filled before moving on
to the next row, if row order recalculation is set.  Column order fills
each column in the range before moving on to the next column.  The
start and increment numbers may be positive or negative.  To fill all
cells with the same value, give an increment of zero.
.IP
.I Example:

util/ss/ss.c
/**********************************************************************
* %M%
* ss 	:	A SpreadSheet Program
*
* Art's Spreadsheet program.          Art Mulder ( art@cs.ualberta.ca )
* University of Alberta, Department of Computing Science.
***********************************************************************
* Main Driver
***********************************************************************
* This program is based (Heavily!) on the well know Public Domain
* spreadsheet program 'sc' Revision 6.19
* May 12/92 -> hacked a bit to bring it up to Rev 6/21 of `sc'
***********************************************************************
* SS Notes & Author List:
*	Feb 1992: sc 6.19 hacked into a new incarnation: 'ss'
*	- currently maintained by Art Mulder ( art@cs.ualberta.ca )
*
* SC Notes & Author List:
*  	original by James Gosling, September 1982
*  	modifications by Mark Weiser and Bruce Israel, 
*  		University of Maryland
*  
*  	More mods Robert Bond, 12/86
*  	More mods by Alan Silverstein, 3-4/88, see list of changes.
*  	Currently supported by sequent!sawmill!buhrt (Jeff Buhrt)
*  	$Revision: 1.1 $
**********************************************************************/
#ifndef lint
  static char Sccsid[] = "%W% %G%";
#endif

/*
 * Include files
 */
#include <stdio.h>
#include <sys/types.h>
#include <signal.h>
#include "curses_stuff.h"
#include <ctype.h>

#ifdef BSD42
# include <strings.h>
#else
# ifndef SYSIII
#   include <string.h>
# endif
#endif

#include "ss.h"
#include "menu.h"
#include "disprange.h"
#include "getinput.h"


/*      Global Variables
 *----------------------------------------------------------------------
 */

/*
 * User I/O variables
 */
char message[MAXSTR];		/* To hold Messages to the user */
char *buff;			/* Pointer to User Input buffer - see
				   the gi_..() functions  */
char *range;			/* Ptr to a range string - see
				   functions that deal with ranges */

/*
 * General Spreadsheet Variables
 */
struct ent ***tbl;		/* data table ref. in vmtbl.c and ATBL() */

  int strow = 0, stcol = 0;
  int currow = 0, curcol = 0;	/* Current location of cell cursor */
  int savedrow, savedcol;
  int maxrow, maxcol;
  int maxrows, maxcols;		/* # cells currently allocated */

char curfile[PATHLEN];		/* Filename of spreadsheet */

/** UNEXPLORED STUFF FOLLOWS **/
/*********************************************************************/
extern	void	startdisp(), stopdisp();

#ifdef SYSV3
  void exit();
#endif

#ifndef SAVENAME
# define	SAVENAME "SC.SAVE" /* file name to use for emergency saves */
#endif /* SAVENAME */


/*
 * Globals defined in sc.h
 */

  int FullUpdate = 0;
  int ClearScreen = 0;	/* don't try to be smart (?? Art M.)*/
  int *fwidth;
  int *precision;
  int *realfmt;
  char *col_hidden;
  char *row_hidden;
  char line[FBUFLEN];		/* Most commands to manipulate the
				   spreadsheet are stored in this
				   variable, and then parsed by a main
				   routine. */
  int changed;
  struct ent *to_fix;
  int modflg;			/* Flag: spreadsheet modified */
  char *mdir;
  int showsc, showsr;	/* Starting cell for highlighted range */
#ifdef RIGHT_CBUG
  int	wasforw	= FALSE;
#endif

/*
 * Function Prototypes
 */
  void	update();
  void	repaint();


  char    revmsg[80];

  int  linelim = -1;

  int  showtop   = 1;	/* Causes current cell value display in top line  */
  int  showcell  = 1;	/* Causes current cell to be highlighted	  */
  int  showrange = 0;	/* Causes ranges to be highlighted		  */
  int  showneed  = 0;	/* Causes cells needing values to be highlighted  */
  int  showexpr  = 0;	/* Causes cell exprs to be displayed, highlighted */

  int  autocalc = 1 ;	/* 1 to calculate after each update */
  int  autolabel = 1;   /* If room, causes label to be created after a define*/
  int  calc_order = BYROWS;
  int  tbl_style = 0;	/* headers for T command output */
  int  rndinfinity = 0;
  int  craction = 0;	/* 1 for down, 2 for right */
  int  rowlimit = -1;
  int  collimit = -1;
#ifdef	SIGWINCH
  int  hitwinch = 0;	/* got a SIGWINCH? */
#endif

  extern int lastmx, lastmy;	/* Screen address of the cursor */
  extern int lastcol, lcols;	/* Spreadsheet Column the cursor was in last */

/* a linked list of free [struct ent]'s, uses .next as the pointer */
  struct ent *freeents = NULL;

  extern	int	seenerr;
  extern	char	*rev;

#ifdef VMS
  int VMS_read_raw = 0;
#endif

char    *progname;
/*********************************************************************/
/** UNEXPLORED STUFF ENDS **/

/*
 * Global Variables 
 */
    int running;	/* Flag for signal a 'quit' request from user */
    int anychanged = FALSE; 

/*      Internal Macros & Data Structures
 *----------------------------------------------------------------------
 */
 

int main (argc, argv)
/*--------------------------------------------------------------------*/
  int 	argc;
  char  **argv;
{
    int     inloop = 1;
    register int   c; 		/* Main input character */
    char    *revi;
/** int	    anychanged = FALSE; **/

/*
 * PROCESS COMMAND LINE OPTIONS
 * 	Keep command line options around until the file is read so the
 * 	command line overrides file options.
 */
    int Mopt = 0;
    int Nopt = 0;
    int Copt = 0; 
    int Ropt = 0;

    int tempx, tempy; 	/* Temp versions of curx, cury */

#if defined(MSDOS)
    if ((revi = strrchr(argv[0], '\\')) != NULL)
#else
# ifdef VMS
    if ((revi = strrchr(argv[0], ']')) != NULL)
# else
    if ((revi = strrchr(argv[0], '/')) != NULL)
# endif
#endif
	progname = revi+1;
    else
	progname = argv[0];

    while (argc > 1 && argv[1][0] == '-') {
	argv++;
	argc--;
    	switch (argv[0][1]) {
#if !defined(VMS) && !defined(MSDOS) && defined(CRYPT_PATH)
	    case 'x':
		    Crypt = 1;
		    break;
#endif
	    case 'm':
		    Mopt = 1;
		    break;
	    case 'c':
		    Copt = 1;
		    break;
	    case 'r':
		    Ropt = 1;
		    break;
	    case 'C':
		    craction = CRCOLS;
		    break;
	    case 'R':
		    craction = CRROWS;
		    break;
	    default:
		    Fprintf(stderr,"%s: unrecognized option: \"%c\"\n",
			progname,argv[0][1]);
		    exit(1);
	}
    }

    *curfile ='\0';

    startdisp();
    signals();

	/* setup the spreadsheet arrays, initscr() will get the screen size */
    if (!growtbl(GROWNEW, 0, 0))
    {	stopdisp();
	exit(1);
    }

/*
 * Build revision message for later use:
 */
    Strcpy (revmsg, progname);
    for (revi = rev; (*revi++) != ':'; );	/* copy after colon */
    Strcat (revmsg, revi);
    revmsg [strlen (revmsg) - 2] = 0;		/* erase last character */

/*  Strcat (revmsg, ":  Type '?' for help, '/' for main menu");  */
    Strcat (revmsg, ":  Type '/' for the main menu."); 
    
/*
 * READ IN SPREADSHEET FILE
 */
    if (argc > 1) {
	Strcpy(curfile,argv[1]);
	readfile (argv[1], 0);
    }

/*
 * APPLY COMMAND LINE OPTIONS
 * 	Command line options override settings stored in spreadsheet
 *	file.  Therefore they are applied after the file is read.
 */
    if (Mopt)
	autocalc = 0;
    if (Copt)
	calc_order = BYCOLS;
    if (Ropt)
	calc_order = BYROWS;

    modflg = 0;
#ifdef VENIX
    setbuf (stdin, NULL);
#endif
    FullUpdate++;

/************
* MAIN LOOP
************/
    while (inloop) { running = 1;
    while (running) {

/* 
 * RECALCULATE THE SPREADSHEET.
 */
/**	if (edistate < 0 && linelim < 0 && autocalc && (changed || FullUpdate))
**/
	if (autocalc && (changed || FullUpdate)) {
	    EvalAll ();
	    if (changed)		/* if EvalAll changed or was before */
		anychanged = TRUE;
	    changed = 0;

	} else if (changed)		/* any cells change? */
	     anychanged = TRUE;

#ifdef	SIGWINCH
	/* got a SIGWINCH? */
	if (hitwinch) {
	    hitwinch = 0;
	    stopdisp();
	    startdisp();
	    FullUpdate++;
	}
#endif
	update(anychanged);
	anychanged = FALSE;

/*
 * GET MAIN ROOT-LEVEL COMMAND/INPUT
 */
	c = nmgetch();			/* Get next input character */
	getyx(stdscr, tempy, tempx);
	ClearMessage;
	Move(tempy, tempx);
/*	(void) fflush (stdout);*/
	seenerr = 0;
	showneed = 0;	/* reset after each update */
	showexpr = 0;

     /* if ((c < ' ') || ( c == DEL )) { how about international here ? PB */
#if pyr
	if ( iscntrl(c) || (c >= 011 && c <= 015) ) {
#else
	if ( ( isascii(c) && iscntrl(c) ) || (c == 020 ) ) {
#endif
	  /* iscntrl broken in OSx4.1 */
	    /*
	     * There seems to be some question about what to do w/ the
	     * iscntrl, some BSD systems are reportedly broken as well.
	     */
            ProcessControlCommands(c);	/* 1) A Control character */

	} else { 	/* NOT a control character, just plain ascii.  */

/*
 * Local Macro: Cannot change locked cells
 */
#define ABORT_IF_LOCKED_CELL 	{if (locked_cell(currow, curcol)) \
				     { beep(); break; } }
		
	    switch (c) {
	      case '/':		/* 2) Enter Command/Menu Mode */
		MainMenu();	
		break;

	      case '?':		/* 3) Invoke Help */
/*   		help(HELP); */
		break;


	      /*
	       * 4) Numeric or Function Entry into cell.
	       *    The case of the '=' differs from the rest in that
	       *    we do not want the '=' put into the user's input
	       *    buffer.
	       */
	      case '=':		/* retained for 'sc' compatibility */
	      case '0': case '1': case '2': case '3': case '4':
	      case '5': case '6': case '7': case '8': case '9':
	      case '-': case '.': case '+':
	      case '@':		/* @ = begin a function */
		ABORT_IF_LOCKED_CELL;
		Message("%s (Value)", v_name(currow, curcol) );

		if ( c != '=' )
		    nmungetch( c ); /* push c back onto input stream */
		
		buff = gi_line_cursor();
		ABORT_AND_BREAK_IF_BUFF_NULL;
		
		Sprintf(line,"let %s = %s", v_name(currow, curcol), buff);
		ClearMessage;
		PROCESS_line;
		break;

	      /*
	       * 5) Label/String Entry into cell.
	       */
	      case '"':		/* 5.a) Centered String */
		ABORT_IF_LOCKED_CELL;
		Message("%s (Label)", v_name(currow, curcol) );
		buff = gi_line_cursor();
		ABORT_AND_BREAK_IF_BUFF_NULL;
		Sprintf (line, "label %s = \"%s\"",
			 v_name(currow, curcol), buff);
		PROCESS_line;
		break;
	      case '>':		/* 5.b) Right Justified String */
		ABORT_IF_LOCKED_CELL;
		Message("%s (Flush-Right Label)", v_name(currow, curcol) );
		buff = gi_line_cursor();
		ABORT_AND_BREAK_IF_BUFF_NULL;
		Sprintf (line, "rightstring %s = \"%s\"",
			 v_name(currow, curcol), buff);
		PROCESS_line;
		break;

	      /* 
	       * The default is to assume that they are starting a
	       * left-justified string.  Make sure that the character
	       * they just entered gets included in the string value.
	       */
	      case '<':		/* 5.c Left Justified String */
	      default:
		ABORT_IF_LOCKED_CELL;
		Message("%s (Flush-Left Label)", v_name(currow, curcol) );

		if ( c != '<' )
		    nmungetch( c ); /* push c back onto input stream */
		
		buff = gi_line_cursor();
		ABORT_AND_BREAK_IF_BUFF_NULL;
		
		Sprintf (line, "leftstring %s = \"%s\"",
			 v_name(currow, curcol), buff);
		PROCESS_line;
		break;

	    }	/* End of Command Switch */
	} /* End of Else */
/*
 * NO MORE Command options.
 */
    }		/* while (running) */
    inloop = modcheck(" before exiting");
    }		/*  while (inloop) */
/******************
* END OF MAIN LOOP
******************/

    stopdisp();
#ifdef VMS	/* Until VMS "fixes" exit we should say 1 here */
    exit(1);
#else
    exit(0);
#endif
    /*NOTREACHED*/

} /* end Main() */

/*********************************************************************/

/* set the calculation order */
void setorder(i)
    int i;
{
    if((i == BYROWS)||(i == BYCOLS))
	calc_order = i;
}

void setauto(i)
    int i;
{
    autocalc = i;
}

void signals()
{
    VOID_OR_INT doquit();
    VOID_OR_INT time_out();
    VOID_OR_INT dump_me();
#ifdef	SIGWINCH
    VOID_OR_INT winchg();
#endif

    Signal(SIGINT, SIG_IGN);
#if !defined(MSDOS)
    Signal(SIGQUIT, dump_me);
    Signal(SIGPIPE, doquit);
    Signal(SIGALRM, time_out);
    Signal(SIGBUS, doquit);
#endif
    Signal(SIGTERM, doquit);
    Signal(SIGFPE, doquit);
#ifdef	SIGWINCH
    Signal(SIGWINCH, winchg);
#endif
}

#ifdef	SIGWINCH
  VOID_OR_INT winchg()
  {	hitwinch++;
	  Signal(SIGWINCH, winchg);
  }
#endif

VOID_OR_INT doquit()
{
    diesave();
    stopdisp();
    exit(1);
}

VOID_OR_INT dump_me()
{
    diesave();
    deraw();
    abort();
}

/* try to save the current spreadsheet if we can */
void diesave()
{   char	path[PATHLEN];

    if (modcheck(" before Spreadsheet dies") == 1)
    {	Sprintf(path, "~/%s", SAVENAME);
	if (writefile(path, 0, 0, maxrow, maxcol) < 0)
	{
	    Sprintf(path, "/tmp/%s", SAVENAME);
	    if (writefile(path, 0, 0, maxrow, maxcol) < 0)
		Message("Couldn't save current spreadsheet, Sorry");
	}
    }
}

/*
 * modcheck()
 *	Check if Spreadsheet has been modified and ask to save
 */
int modcheck(endstr)
  char *endstr;		/* message to user */
{
    if (modflg && curfile[0]) {
	int	yn_ans;
	char	lin[100];

	Sprintf (lin,"File \"%s\" is modified, save%s? ",
		curfile,endstr);
	if ((yn_ans = yn_ask(lin)) < 0)
	    return(1);
	else if (yn_ans == 1) {
	    if (writefile(curfile, 0, 0, maxrow, maxcol) < 0)
 		return (1);
	}
    } else if (modflg) {
	int	yn_ans;

	if ((yn_ans = yn_ask("Do you want a chance to save the data? ")) < 0)
	    return(1);
	else
	    return(yn_ans);
    }
    return(0);
} /* modcheck() */

/* Returns 1 if cell is locked, 0 otherwise */
int locked_cell (r, c)
    int r, c;
{
    struct ent *p = *ATBL(tbl, r, c);
    if (p && (p->flags & is_locked)) {
	Message("Cell %s%d is locked", coltoa(c), r) ;
	return(1);
    }
    return(0);
}

/* Check if area contains locked cells */
int any_locked_cells(r1, c1, r2, c2)
    int r1, c1, r2, c2 ;
{
    int r, c;
    struct ent *p ;

    for (r=r1; r<=r2; r++)
	for (c=c1; c<=c2; c++) {
	    p = *ATBL(tbl, r, c);
	    if (p && (p->flags&is_locked))
		return(1);
	}
    return(0);
}

/* 
 * return a pointer to a cell's [struct ent *], creating if needed 
 */
struct ent * lookat(row,col)
    int	row, col;
{
    register struct ent **pp;

    checkbounds(&row, &col);
    pp = ATBL(tbl, row, col);
    if (*pp == (struct ent *)0) {
        if (freeents != NULL)
	{	*pp = freeents;
		freeents = freeents->next;
	}
	else
		*pp = (struct ent *) Malloc((unsigned)sizeof(struct ent));
	if (row>maxrow) maxrow = row;
	if (col>maxcol) maxcol = col;
	(*pp)->label = (char *)0;
	(*pp)->row = row;
	(*pp)->col = col;
	(*pp)->flags = 0;
	(*pp)->expr = (struct enode *)0;
	(*pp)->v = (double) 0.0;
	(*pp)->format = (char *)0;
	(*pp)->cellerror = CELLOK;
	(*pp)->next = NULL;
    }
    return *pp;
}

/*
 * This structure is used to keep ent structs around before they
 * are deleted to allow the sync_refs routine a chance to fix the
 * variable references.
 * We also use it as a last-deleted buffer for the 'p' command.
 */
void free_ent(p)
    register struct ent *p;
{
    p->next = to_fix;
    to_fix = p;
    p->flags |= is_deleted;
    p->flags &= ~is_locked;
}

/* free deleted cells */
void flush_saved()
{
    register struct ent *p;
    register struct ent *q;

    if (!(p = to_fix))
	return;
    while (p) {
	clearent(p);
	q = p->next;
	p->next = freeents;	/* put this ent on the front of freeents */
	freeents = p;
	p = q;
    }
    to_fix = NULL;
}
/**********************************************************************
*	END
**********************************************************************/
